plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -0,0 +1,420 @@
+  private OrientVertex findAsset(final StorageTx tx,
+                                 final MavenPath mavenPath,
+                                 final OrientVertex bucket)
+  {
+    final String assetKeyName =
+        StorageFacet.P_ATTRIBUTES + "." + getRepository().getFormat().getValue() + "." + P_ASSET_KEY;
+    return tx.findAssetWithProperty(assetKeyName, getAssetKey(mavenPath), bucket);
+  }
+
+  private NestedAttributesMap getAttributes(final StorageTx tx, final OrientVertex vertex) {
+    return tx.getAttributes(vertex);
+  }
+
+  private NestedAttributesMap getFormatAttributes(final StorageTx tx, final OrientVertex vertex) {
+    return getAttributes(tx, vertex).child(getRepository().getFormat().getValue());
+  }
+
+  /**
+   * Determines or confirms the content type for the content, or throws {@link InvalidContentException} if it cannot.
+   */
+  @Nonnull
+  private String determineContentType(final MavenPath mavenPath,
+                                      final Supplier<InputStream> inputStreamSupplier,
+                                      final String declaredContentType)
+      throws IOException
+  {
+    String contentType = declaredContentType;
+
+    if (contentType == null) {
+      log.trace("Content PUT to {} has no content type.", mavenPath);
+      try (InputStream is = inputStreamSupplier.get()) {
+        contentType = mimeSupport.detectMimeType(is, mavenPath.getPath());
+      }
+      log.trace("Mime support implies content type {}", contentType);
+
+      if (contentType == null && strictContentTypeValidation) {
+        throw new InvalidContentException(String.format("Content type could not be determined."));
+      }
+    }
+    else {
+      try (InputStream is = inputStreamSupplier.get()) {
+        final List<String> types = mimeSupport.detectMimeTypes(is, mavenPath.getPath());
+        if (!types.isEmpty() && !types.contains(contentType)) {
+          log.debug("Discovered content type {} ", types.get(0));
+          if (strictContentTypeValidation) {
+            throw new InvalidContentException(
+                String.format("Declared content type %s, but declared %s.", contentType, types.get(0)));
+          }
+        }
+      }
+    }
+    return contentType;
+  }
+
+  private StorageFacet getStorage() {
+    return getRepository().facet(StorageFacet.class);
+  }
+
+  private BlobRef getBlobRef(final MavenPath mavenPath, final OrientVertex asset) {
+    String blobRefStr = asset.getProperty(StorageFacet.P_BLOB_REF);
+    checkState(blobRefStr != null, "asset of component at path %s has missing blob reference",
+        mavenPath.getPath());
+    return BlobRef.parse(blobRefStr);
+  }
+
+  /**
+   * Creates a payload out of a Blob, for content originating from Blobs.
+   */
+  private Payload marshall(final StorageTx tx, final OrientVertex asset, final Blob blob) {
+    final String contentType = asset.getProperty(StorageFacet.P_CONTENT_TYPE);
+    return new BlobPayload(blob, contentType);
```

Comment:
hrm... I don't think marshall() is a good name, its technically not marshalling anything here.  Its just adapting a blob to a payload